src/java.base/share/classes/java/lang/StringUTF16.java

```diff
@@ -158,30 +177,210 @@ public static byte[] toBytes(char[] value, int off, int len) {
-        byte[] ret = new byte[len];
-        if (compress(val, off, ret, 0, len) == len) {
-            return ret;
+    /**
+     * {@return Compress the internal byte array (containing UTF16) into a compact strings byte array}
+     * If all the chars are LATIN1, it returns an array with len == count,
+     * otherwise, it contains UTF16 characters.
+     * <p>
+     * Refer to the description of the algorithm in {@link #compress(char[], int, int)}.
+     *
+     * @param val   a byte array with UTF16 coding
+     * @param off   starting offset
+     * @param count count of chars to be compressed, {@code count} > 0
+     */
+    public static byte[] compress(final byte[] val, final int off, final int count) {
+        byte[] latin1 = new byte[count];
+        int ndx = compress(val, off, latin1, 0, count);
+        if (ndx != count) {// Switch to UTF16
+            byte[] utf16 = Arrays.copyOfRange(val, off << 1, newBytesLength(off + count));
+            // If the original character that was found to be non-latin1 is latin1 in the copy
+            // try to make a latin1 string from the copy
+            if (getChar(utf16, ndx) > 0xff
+                    || compress(utf16, 0, latin1, 0, count) != count) {
+                return utf16;
+            }
         }
-        return null;
+        return latin1;     // latin1 success
+    }
+
+    /**
+     * {@return compress the code points into a compact strings byte array}
+     * If all the chars are LATIN1, returns an array with len == count.
+     * If not, a new byte array is allocated and code points converted to UTF16.
+     * The algorithm is similar to that of {@link #compress(char[], int, int)}.
+     * <p>
+     * The resulting encoding is attempted in several steps:
+     * <UL>
+     *     <LI>If no non-latin1 characters are found, the encoding is latin1</LI>
+     *     <LI>If an estimate of the number of characters needed to represent the codepoints is
+     *         equal to the string length, they are all BMP with at least 1 UTF16 character
+     *         and are copied to the result. </LI>
+     *     <LI>The extractCodePoints method is called to carefully expand surrogates. </LI>
+     * </UL>
+     *
+     * @param val   an int array of code points
+     * @param off   starting offset
+     * @param count length of code points to be compressed, length > 0
+     */
+    public static byte[] compress(final int[] val, int off, final int count) {
+        // Optimistically copy all latin1 code points to the destination
+        byte[] latin1 = new byte[count];
+        final int end = off + count;
+        for (int ndx = 0; ndx < count; ndx++, off++) {
+            int cp = val[off];
+            if (cp >= 0 && cp <= 0xff) {
+                latin1[ndx] = (byte)cp;
+            } else {
+                // Pass 1: Compute precise size of char[]; see extractCodePoints for caveat
+                int estSize = ndx + computeCodePointSize(val, off, end);
```

Comment:
Interesting idea, but it might mean that if the codepoint val[i] was modified it could result in a cp that did not (ever) exist in the input; creating a value out of thin air.  The high bits would be from the first read of val[i] and the low bits from the 2nd read. The code in extractCodepoints could be simpler and computeCodePointSize just a little more comples. Creating values out of thin air is usually bad and could have (different) unexpected results in the app.

The additional writes to the latin1 array would also slow down the normal case of computing the size whether or not the input was modified.
On that basis, I'd keep the current approach to resizing only if needed.